src/daft-scan/src/lib.rs

```diff
@@ -1,93 +1,318 @@
+        )
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
 pub struct PartitionField {
     field: Field,
     source_field: Option<Field>,
     transform: Option<Expr>,
 }
 
-pub trait ScanOperator: Send + Display {
+pub trait ScanOperator: Send + Sync + Display + Debug + DynHash {
     fn schema(&self) -> SchemaRef;
     fn partitioning_keys(&self) -> &[PartitionField];
-    fn num_partitions(&self) -> DaftResult<usize>;
+    // fn statistics(&self) -> &TableStatistics;
+    // fn clustering_spec(&self) -> &ClusteringSpec;
+
+    fn can_absorb_filter(&self) -> bool;
+    fn can_absorb_select(&self) -> bool;
+    fn can_absorb_limit(&self) -> bool;
+    fn to_scan_tasks(
+        &self,
+        pushdowns: Pushdowns,
+    ) -> DaftResult<Box<dyn Iterator<Item = DaftResult<ScanTask>>>>;
+}
+
+// The following implements PartialEq, Eq, and Hash for trait objects.
+
+pub trait DynEq: Any {
+    fn as_any(&self) -> &dyn Any;
+    fn dyn_eq(&self, that: &dyn DynEq) -> bool;
+}
+
+impl<T: Any + PartialEq<Self>> DynEq for T {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+    fn dyn_eq(&self, that: &dyn DynEq) -> bool {
+        if let Some(that) = that.as_any().downcast_ref::<Self>() {
+            self == that
+        } else {
+            false
+        }
+    }
+}
+
+pub trait DynHash: DynEq {
+    fn dyn_hash(&self, hasher: &mut dyn Hasher);
 
-    // also returns a bool to indicate if the scan operator can "absorb" the predicate
-    fn filter(self: Box<Self>, predicate: &Expr) -> DaftResult<(bool, ScanOperatorRef)>;
-    fn select(self: Box<Self>, columns: &[&str]) -> DaftResult<ScanOperatorRef>;
-    fn limit(self: Box<Self>, num: usize) -> DaftResult<ScanOperatorRef>;
-    fn to_scan_tasks(self: Box<Self>)
-        -> DaftResult<Box<dyn Iterator<Item = DaftResult<ScanTask>>>>;
+    fn as_dyn_eq(&self) -> &dyn DynEq;
 }
 
-pub type ScanOperatorRef = Box<dyn ScanOperator>;
+impl<H: Hash + DynEq> DynHash for H {
+    fn dyn_hash(&self, mut hasher: &mut dyn Hasher) {
+        H::hash(self, &mut hasher)
+    }
+
+    fn as_dyn_eq(&self) -> &dyn DynEq {
+        self
+    }
+}
+
+impl PartialEq for dyn ScanOperator + '_ {
+    fn eq(&self, that: &dyn ScanOperator) -> bool {
+        self.dyn_eq(that.as_dyn_eq())
+    }
+}
+
+impl PartialEq<dyn ScanOperator> for Box<dyn ScanOperator + '_> {
+    fn eq(&self, that: &dyn ScanOperator) -> bool {
+        self.dyn_eq(that.as_dyn_eq())
+    }
+}
+
+impl PartialEq<dyn ScanOperator> for Arc<dyn ScanOperator + '_> {
+    fn eq(&self, that: &dyn ScanOperator) -> bool {
+        self.dyn_eq(that.as_dyn_eq())
+    }
+}
+
+impl Eq for dyn ScanOperator + '_ {}
+
+impl Hash for dyn ScanOperator + '_ {
+    fn hash<H: Hasher>(&self, hasher: &mut H) {
+        self.dyn_hash(hasher)
+    }
+}
+
+pub type ScanOperatorRef = Arc<dyn ScanOperator>;
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct ScanExternalInfo {
+    pub scan_op: Arc<dyn ScanOperator>,
```

Comment:
I was thinking for equality and hashing for this, we could just use the address of the scan_op so scan operators dont have to implement hashing or equals

Reply:
See: https://github.com/Eventual-Inc/Daft/pull/1562